Added "via" and "on" to prompt variables

mawburn opened the pull request:
I didn't like the verboseness of "via" and "on" so I added them to options so I could set them to emoji. `PROMPT` seems like the best place to have added it.

denysdovhan commented:
Yeah, I've thought about it, but I think it would be better to give users ability to disable all `in`, `on`, `at`, `via` prefixes. 

I think it's better to add `SPACESHIP_[SECTION]_PREFIX` variables for each of sections. What do you think?

mawburn commented:
Yeah, I just started using your theme and missed `in` and `at`... but had I used SSH with this I probably have added those too. 😋

I agree that many options would warrant it's own config section. But with just the 2 I added, it would have been overkill to make a whole new config group.

denysdovhan commented:
Just to make things clear, I suggest you not to create new section, but add `_PREFIX` option to each of them, like:

* `SPACESHIP_GIT_PREFIX`
* `SPACESHIP_VENV_PREFIX`
* `SPACESHIP_RUBY_PREFIX`
* and so on

Also, I'm gonna merge #40, so you will have to care about it.

mawburn commented:
Ah ok, got it. What about a PREFIX scope (`SPACESHIP_PREFIX_GIT`), so that way the user could define them all at once or turn them on/off with something like `SPACESHIP_PREFIX_SHOW`? 

```

denysdovhan commented:
Looks good. Where are you gonna put `at` and `in` prefixes?

mawburn commented:
Not sure. Maybe directly on `SPACESHIP_PREFIX_SSH_HOST` and `SPACESHIP_PREFIX_SSH_USER`?

denysdovhan commented:
Actually, `HOST` and `USER` parts is only about SSH. For example, `USER` is shown when you're `root`, so it would be better to name those variables as here:

* `SPACESHIP_PREFIX_HOST`
* `SPACESHIP_PREFIX_USER`
* `SPACESHIP_PREFIX_GIT`
* `SPACESHIP_PREFIX_NVM`
* `SPACESHIP_PREFIX_RUBY`
* and so on

mawburn commented:
K. I can do that. That looks better. I didn't like the `_SSH_` either.  Wasn't able to get around to it last night, but I will sometime this weekend.

denysdovhan commented:
@mawburn yeah, sure. But don't forget about it. I'm gonna merge #40 after this PR.

mawburn commented:
Thought I squashed both of those. 

Anyway, seems to work. Please verify the new settings. I wasn't sure what to name the SSH stuff once I got in there, so I named them:

>`SPACESHIP_PREFIX_MACHINE`, `SPACESHIP_PREFIX_SSH_DIR`

I couldn't seem to get my Ruby or Venv working, but I work in Node, so I wasn't sure if it was an issue with the theme or the way I had it installed.

denysdovhan commented on spaceship.zsh:
```diff
@@ -16,19 +16,13 @@ SPACESHIP_PROMPT_TRUNC="${SPACESHIP_PROMPT_TRUNC:-3}"
-SPACESHIP_PREFIX_GIT="${SPACESHIP_PREFIX_GIT:-on}"
-SPACESHIP_PREFIX_HOST="${SPACESHIP_PREFIX_HOST:-at}"
-SPACESHIP_PREFIX_USER="${SPACESHIP_PREFIX_USER:-in}"
+SPACESHIP_PREFIX_MACHINE="${SPACESHIP_PREFIX_MACHINE:-at}"
```
I think `SPACESHIP_PREFIX_HOST` and `SPACESHIP_PREFIX_DIR` would be much more consistence and obvious names for these variables. Please rename here and update the docs.

denysdovhan commented on README.md:
```diff
@@ -106,17 +106,25 @@ Now you have ability to disable elements of Spaceship. All options must be overr
 | `SPACESHIP_PROMPT_ADD_NEWLINE` | `true` | Adds a newline character before each prompt line |
 | `SPACESHIP_PROMPT_TRUNC` | `3` | Number of folders of cwd to show in prompt, 0 to show all |
 
+### Prefixes
```
Let's put all prefix-related variables in this section. That will be more predictable. Probably user doesn't expect to see `_PREFIX_` variable among `_GIT_` or `_NVM_` ones, right?

denysdovhan commented on spaceship.zsh:
```diff
@@ -16,19 +16,13 @@ SPACESHIP_PROMPT_TRUNC="${SPACESHIP_PROMPT_TRUNC:-3}"
 
 # GIT
 SPACESHIP_GIT_SHOW="${SPACESHIP_GIT_SHOW:-true}"
+SPACESHIP_GIT_PREFIX="${SPACESHIP_GIT_PREFIX:-on}"
```
Here the same as above: put all related to `_PREFIX_` variables in corresponding section.

denysdovhan commented on spaceship.zsh:
```diff
@@ -70,12 +74,24 @@ spaceship_user() {
   if [[ -n $SSH_CONNECTION ]]; then
     echo -n "$(spaceship_user)"
-    echo -n " %B${SPACESHIP_PREFIX_USER}%b "
+
```
Actually, this code can be rewritten simpler. You can simplify this:

```sh
if [[ ${SPACESHIP_PREFIX_SHOW} == true ]]; then 
  echo -n " %B${SPACESHIP_PREFIX_SSH_DIR}%b "
fi
```

to this:

```sh
[[ ${SPACESHIP_PREFIX_SHOW} == true ]] && echo -n " %B${SPACESHIP_PREFIX_SSH_DIR}%b "
```

I'm sure this small change is gonna make code more clean.

mawburn replied:
I was wondering about that. I saw that elsewhere in your code, but I couldn't find any documentation supporting it. Thanks.

mawburn commented:
Updated and cleaned up the variables. 

I feel a little silly about this, but it turned out the default variable wasn't being set the way I intended so I had to initialize it after zsh read in the config using a `set_prefixes` function.

denysdovhan commented on spaceship.zsh:
```diff
@@ -33,29 +33,29 @@ SPACESHIP_GIT_UNPUSHED="${SPACESHIP_GIT_UNPUSHED:-⇡}"
 
+# Prefixes
+# Initializes prefix variables
+spaceship_prefixes() {
```
Why did you put this variables in `spaceship_prefixes` function?

mawburn replied:
Because it needs to run after the .zshrc is initialized, so that it can initialize them all to a new variable. If you move the variables to the top with the others, they stay set to `via`, no matter what you have set in your .zshrc.

mawburn replied:
Which looks like it breaks the explicit sets. 

I'm not sure how to get around this. The best way, maybe, is to take out the default initialization variable and set them all explicitly.

denysdovhan replied:
Okay, I'm going to check it. Give me some time.

denysdovhan commented:
Merged! Thank you for contributing!

denysdovhan commented:
<blockquote class="twitter-tweet" data-partner="tweetdeck"><p lang="en" dir="ltr">spaceship-zsh-theme became even more customizable! Now you can set your own prefixes! <a href="https://t.co/fX0O87jk2y">https://t.co/fX0O87jk2y</a><br>Thanks to <a href="https://twitter.com/mawburn_">@mawburn_</a> ! <a href="https://t.co/TsQZQAO8zp">pic.twitter.com/TsQZQAO8zp</a></p>&mdash; Denys Dovhan (@denysdovhan) <a href="https://twitter.com/denysdovhan/status/801071711680626688">November 22, 2016</a></blockquote>

cc/ @mawburn
